### ISSUE TITLE
Cannot Resolve Components In A Template Collocation App

### ISSUE THREAD

**Issue by chadhietala** [embroider-build/embroider#357]
With template collocation, an application should be able to place their component template next to their class (if one exists). In Octane this is the new default.

**Steps To Reproduce**

1. `ember new ember-quickstart -b @ember/octane-app-blueprint`
2. `ember g component foo`
3. Opt-in to using embroider [as specified in the README](https://github.com/embroider-build/embroider#how-to-try-it)
3. Place `<Foo />` in application.hbs
4. `ember s`
5. Go to the app and you will get the following error:

```
vendor.js:1189 Uncaught (in promise) Error: Failed to create an instance of 'component:navigation'. Most likely an improperly defined class or an invalid module export.
    at FactoryManager.create (vendor.js:1189)
    at CurlyComponentManager.create (vendor.js:7458)
    at Object.evaluate (vendor.js:49188)
    at AppendOpcodes.evaluate (vendor.js:47760)
    at LowLevelVM.evaluateSyscall (vendor.js:50982)
    at LowLevelVM.evaluateInner (vendor.js:50928)
    at LowLevelVM.evaluateOuter (vendor.js:50920)
    at VM.next (vendor.js:52986)
    at TemplateIteratorImpl.next (vendor.js:53022)
    at RootState.render (vendor.js:7903)
```

**Comment by NullVoxPopuli** [comment]
I got a similar error, but with nested components: 
![image](https://user-images.githubusercontent.com/199018/69757804-eb780a00-112b-11ea-9c87-c838e67e334a.png)
![image](https://user-images.githubusercontent.com/199018/69757820-f2068180-112b-11ea-873e-2ee97ba930be.png)

**Comment by lifeart** [comment]
This issue can be fixed adding new paths here: https://github.com/lifeart/embroider/blob/master/packages/compat/src/resolver.ts#L304

**Comment by ef4** [comment]
Template co-location has a different implementation on the Ember side, so I think it may not be enough to just find the right template and register it with the container as we do now. Instead we can emit the new `setComponentTemplate` that was created for this purpose.

In fact, in any app that has a new enough ember, we should probably switch to using `setComponentTemplate` for all components that have templates. It's better aligned with how Embroider works anyway. It expresses a direct link between the JS and the HBS, instead of registering a globally-visible path on the resolver that just happens to match them together via conventions.

**Comment by rwjblue** [comment]
> In fact, in any app that has a new enough ember, we should probably switch to using setComponentTemplate for all components that have templates. 

Yes, completely agree.

**Comment by NullVoxPopuli** [comment]
Just got passed the typescript error from #352,

my current error is: 
```
This problem is likely caused by another plugin injecting
"QUnit" without registering it in the scope tracker. If you are the author
 of that plugin, please use "scope.registerDeclaration(declarationPath)".
(node:6890) UnhandledPromiseRejectionWarning: ModuleBuildError: Module build failed (from /home/lprestonsegoiii/Development/NullVoxPopuli/emberclear/packages/frontend/node_modules/@embroider/webpack/src/webpack-hbs-loader.js):
Error: Missing component Chat::ChatHistory::NewMessages in /tmp/embroider/25570d/pods/components/pod/chat/chat-history/template.hbs
    at CompatResolver.dependenciesOf (/home/lprestonsegoiii/Development/NullVoxPopuli/emberclear/packages/frontend/node_modules/@embroider/compat/src/resolver.js:184:31)
    at TemplateCompiler.precompile (/home/lprestonsegoiii/Development/NullVoxPopuli/emberclear/packages/frontend/node_modules/@embroider/core/src/template-compiler.js:246:49)
```
embroider 0.10.0, yarn outdated shows nothing outdated :)

**Comment by NullVoxPopuli** [comment]
so what should be done here? I think maybe this is related to: https://github.com/embroider-build/embroider/issues/439

**Comment by ef4** [comment]
Support for template co-location was implemented back in 0.12.0, so this issue should have been closed.

We can continue discussion on #439 about your specific bug.
